lektor/imagetools.py

```diff
@@ -449,40 +507,74 @@ def process_image(ctx, source_image, dst_filename, width, height=None,
     portable_popen(cmdline).wait()
 
 
-def make_thumbnail(ctx, source_image, source_url_path, width, height=None,
-                   crop=False, quality=None):
+def make_thumbnail(ctx, source_image, source_url_path,
+                   width=None, height=None, mode=ThumbnailMode.DEFAULT,
+                   upscale=None, quality=None):
     """Helper method that can create thumbnails from within the build process
     of an artifact.
     """
+    if width is None and height is None:
+        raise ValueError("Must specify at least one of width or height.")
+
+    if upscale is None:
+        upscale = {
+            ThumbnailMode.FIT: False,
+            ThumbnailMode.CROP: True,
+            ThumbnailMode.STRETCH: True,
+        }[mode]
+
+    # fallback to "fit" in case of erroneous arguments (but preserve
+    # the `upscale` defaults as defined above).
+    if mode != ThumbnailMode.FIT and (width is None or height is None):
+        warnings.warn(
+            '"%s" mode requires both `width` and `height` to be defined. '
+            'Falling back to "fit" mode.`' % mode.label
+        )
+        mode = ThumbnailMode.FIT
```

Comment:
@xlotlu From previous comment about this, this should be breaking here (as of the latest breaking commit), and in the final backwards-compatible commit it should look like this to be non-breaking.